File: Julia/resnet-identity-block.jl
relu(x) = max.(0, x)

mutable struct IdentityBlock
    channels::Int
    W1
    W2
end

function IdentityBlock(channels::Int)
    W1 = randn(channels, channels) .* 0.01
    W2 = randn(channels, channels) .* 0.01
    IdentityBlock(channels, W1, W2)
end

function forward(block::IdentityBlock, x)
    identity = x
    out = relu(x * block.W1)
    out = out * block.W2
    out .+ identity
end

File: R/resnet-identity-block.R
relu <- function(x) {
  pmax(0, x)
}

IdentityBlock <- setRefClass(
  "IdentityBlock",
  fields = list(
    channels = "numeric",
    W1 = "matrix",
    W2 = "matrix"
  ),
  methods = list(
    initialize = function(channels) {
      channels <<- channels
      W1 <<- matrix(rnorm(channels * channels, sd = 0.01), nrow = channels, ncol = channels)
      W2 <<- matrix(rnorm(channels * channels, sd = 0.01), nrow = channels, ncol = channels)
    },
    forward = function(x) {
      identity <- x
      out <- relu(x %*% W1)
      out <- out %*% W2
      out + identity
    }
  )
)
